File: Nori/Store/HistoryStore.swift
import Foundation
import Observation
import OSLog
import SwiftData

/// The observable view of the persisted history, backed by SwiftData.
///
/// All mutation goes through here so ordering, de-duplication, the size limit and expiry
/// are enforced in one place. Views only ever see `[ClipRow]` values.
@MainActor
@Observable
final class HistoryStore {
    /// Newest first. Rebuilt on every mutation.
    private(set) var rows: [ClipRow] = []
    /// Bumped on every mutation.
    private(set) var version = 0
    var maxItems: Int = 500 {
        didSet {
            guard maxItems != oldValue else { return }
            if enforceLimit() { save() } else { publish() }
        }
    }
    /// 0 = never.
    var expireAfterDays: Int = 0

    /// A deleted item that can still be brought back with ⌘Z.
    struct UndoRecord {
        let draft: ClipDraft
        let firstCopiedAt: Date
        let lastCopiedAt: Date
        let copyCount: Int
        let pinnedAt: Date?
    }

    private var items: [ClipItem] = []
    private let context: ModelContext
    private let logger = Logger(subsystem: "io.github.hocky0301.Nori", category: "history")

    init(context: ModelContext) {
        self.context = context
    }

    var count: Int { items.count }
    var pinnedCount: Int { items.filter(\.isPinned).count }
    var totalBytes: Int { items.reduce(0) { $0 + $1.byteCount } }

    func load() {
        var descriptor = FetchDescriptor<ClipItem>(sortBy: [SortDescriptor(\.lastCopiedAt, order: .reverse)])
        descriptor.includePendingChanges = true
        do {
            items = try context.fetch(descriptor)
        } catch {
            logger.error("fetch failed: \(error.localizedDescription)")
            items = []
        }
        expireOldItems()
        enforceLimit()
        save()
    }

    func item(id: UUID) -> ClipItem? {
        items.first { $0.id == id }
    }

    func row(id: UUID) -> ClipRow? {
        rows.first { $0.id == id }
    }

    /// Insert a new capture, or refresh the existing entry with identical content.
    @discardableResult
    func ingest(_ draft: ClipDraft, now: Date = .now) -> UUID {
        if let existing = items.first(where: { $0.contentHash == draft.contentHash }) {
            existing.lastCopiedAt = now
            existing.copyCount += 1
            if let bundle = draft.sourceBundleID {
                existing.sourceBundleID = bundle
                existing.sourceAppName = draft.sourceAppName
            }
            moveToTop(existing)
            save()
            return existing.id
        }

        let item = ClipItem(draft: draft, now: now)
        context.insert(item)
        attach(draft.contents, to: item)
        insertByRecency(item)
        enforceLimit()
        save()
        return item.id
    }

    /// Called when the user pasted an item through Nori: it becomes the most recent copy again.
    func touch(id: UUID, now: Date = .now) {
        guard let item = item(id: id) else { return }
        item.lastCopiedAt = now
        item.copyCount += 1
        moveToTop(item)
        save()
    }

    /// Remove an item; the returned record restores it with dates, count and pin intact.
    @discardableResult
    func delete(id: UUID) -> UndoRecord? {
        guard let item = item(id: id) else { return nil }
        var draft = ClipDraft(kind: item.kind, contents: contents(id: id), contentHash: item.contentHash,
                              isFromUniversalClipboard: item.isFromUniversalClipboard)
        draft.title = item.title
        draft.searchText = item.searchText
        draft.sourceBundleID = item.sourceBundleID
        draft.sourceAppName = item.sourceAppName
        draft.isRichText = item.isRichText
        draft.isTruncated = item.isTruncated
        draft.linkURL = item.linkURL
        draft.colorHex = item.colorHex
        draft.fileURLs = item.fileURLs
        draft.imagePixelSize = item.imagePixelSize
        draft.thumbnail = item.thumbnail
        draft.characterCount = item.characterCount
        draft.lineCount = item.lineCount
        let record = UndoRecord(draft: draft, firstCopiedAt: item.firstCopiedAt, lastCopiedAt: item.lastCopiedAt,
                                copyCount: item.copyCount, pinnedAt: item.pinnedAt)
        items.removeAll { $0.id == id }
        context.delete(item)
        save()
        return record
    }

    /// Bring back a deleted item exactly where it was. If the same content was copied again in
    /// the meantime, the counts merge into that row instead of duplicating it.
    @discardableResult
    func restore(_ record: UndoRecord) -> UUID {
        if let existing = items.first(where: { $0.contentHash == record.draft.contentHash }) {
            existing.copyCount += record.copyCount
            existing.firstCopiedAt = min(existing.firstCopiedAt, record.firstCopiedAt)
            if existing.pinnedAt == nil { existing.pinnedAt = record.pinnedAt }
            save()
            return existing.id
        }
        let item = ClipItem(draft: record.draft, now: record.firstCopiedAt)
        item.lastCopiedAt = record.lastCopiedAt
        item.copyCount = record.copyCount
        item.pinnedAt = record.pinnedAt
        context.insert(item)
        attach(record.draft.contents, to: item)
        let index = items.firstIndex { $0.lastCopiedAt < record.lastCopiedAt } ?? items.count
        items.insert(item, at: index)
        save()
        return item.id
    }

    /// Remove history. Pinned items survive unless `includingPinned` is set.
    @discardableResult
    func clear(includingPinned: Bool) -> Int {
        let doomed = includingPinned ? items : items.filter { !$0.isPinned }
        let ids = Set(doomed.map(\.id))
        items.removeAll { ids.contains($0.id) }
        doomed.forEach(context.delete)
        save()
        return doomed.count
    }

    /// Unpinning never evicts the item just unpinned; the cap is applied again on the next capture.
    func togglePin(id: UUID, now: Date = .now) {
        guard let item = item(id: id) else { return }
        item.pinnedAt = item.isPinned ? nil : now
        save()
    }

    /// Update the searchable text of an item (used when OCR finishes for an image).
    func updateSearchText(id: UUID, text: String) {
        guard let item = item(id: id) else { return }
        item.searchText = String(text.prefix(ClipDraft.maxSearchTextLength))
        save()
    }

    // MARK: Full content (on demand)

    // Payloads are read through a throwaway context: the external-storage blobs are cached on
    // the managed object that loaded them, and `items` keeps its objects alive for the app's
    // lifetime, so reading through them would keep every pasted or previewed image resident.

    func contents(id: UUID) -> [ClipDraft.Content] {
        withPayload(id: id) { $0.draftContents } ?? []
    }

    func plainText(id: UUID) -> String? {
        withPayload(id: id) { $0.plainText } ?? nil
    }

    func imageData(id: UUID) -> Data? {
        withPayload(id: id) { $0.imageData } ?? nil
    }

    private func withPayload<T>(id: UUID, _ read: (ClipItem) -> T) -> T? {
        guard item(id: id) != nil else { return nil }
        let scratch = ModelContext(context.container)
        scratch.autosaveEnabled = false
        var descriptor = FetchDescriptor<ClipItem>(predicate: #Predicate { $0.id == id })
        descriptor.fetchLimit = 1
        guard let fetched = try? scratch.fetch(descriptor).first else { return nil }
        return read(fetched)
    }

    // MARK: Maintenance

    /// Drop unpinned items older than `expireAfterDays`.
    func expireOldItems(now: Date = .now) {
        guard expireAfterDays > 0 else { return }
        let cutoff = now.addingTimeInterval(-Double(expireAfterDays) * 86_400)
        let doomed = items.filter { !$0.isPinned && $0.lastCopiedAt < cutoff }
        guard !doomed.isEmpty else { return }
        let ids = Set(doomed.map(\.id))
        items.removeAll { ids.contains($0.id) }
        doomed.forEach(context.delete)
        save()
    }

    // MARK: - Private

    private func attach(_ contents: [ClipDraft.Content], to item: ClipItem) {
        for content in contents {
            let row = ClipContent(type: content.type, data: content.data)
            context.insert(row)
            row.item = item
        }
    }

    private func moveToTop(_ item: ClipItem) {
        items.removeAll { $0.id == item.id }
        insertByRecency(item)
    }

    /// Keep `items` sorted by `lastCopiedAt` (newest first) even when captures arrive with earlier timestamps.
    private func insertByRecency(_ item: ClipItem) {
        let index = items.firstIndex { $0.lastCopiedAt <= item.lastCopiedAt } ?? items.count
        items.insert(item, at: index)
    }

    /// Returns true when something was evicted (the caller saves).
    @discardableResult
    private func enforceLimit() -> Bool {
        let unpinned = items.filter { !$0.isPinned }
        guard unpinned.count > maxItems else { return false }
        let overflow = unpinned[maxItems...]
        let overflowIDs = Set(overflow.map(\.id))
        items.removeAll { overflowIDs.contains($0.id) }
        overflow.forEach(context.delete)
        return true
    }

    private func save() {
        do {
            try context.save()
        } catch {
            logger.error("save failed: \(error.localizedDescription)")
        }
        publish()
    }

    private func publish() {
        rows = items.map { $0.makeRow() }
        version += 1
    }
}

File: windows/Nori.Core/HistoryStore.cs
namespace Nori.Core;

/// <summary>One persisted clip: the row metadata plus its representations (loaded on demand by the repository).</summary>
public sealed record StoredClip
{
    public required Guid Id { get; init; }
    public required ClipKind Kind { get; init; }
    public required string Title { get; init; }
    public required string SearchText { get; init; }
    public required string ContentHash { get; init; }
    public string? SourceApp { get; init; }
    public string? SourceAppName { get; init; }
    public string? SourceAppPath { get; init; }
    public required DateTimeOffset FirstCopiedAt { get; init; }
    public required DateTimeOffset LastCopiedAt { get; init; }
    public int CopyCount { get; init; } = 1;
    public DateTimeOffset? PinnedAt { get; init; }
    public bool IsRichText { get; init; }
    public bool IsTruncated { get; init; }
    public int LineCount { get; init; }
    public int CharacterCount { get; init; }
    public int ByteCount { get; init; }
    public Uri? LinkUrl { get; init; }
    public string? ColorHex { get; init; }
    public IReadOnlyList<string> FilePaths { get; init; } = [];
    public (int Width, int Height)? ImagePixelSize { get; init; }
    public byte[]? Thumbnail { get; init; }

    public bool IsPinned => PinnedAt is not null;

    public ClipRow ToRow() => new()
    {
        Id = Id,
        Source = RowSource.History,
        Kind = Kind,
        Title = Title,
        SearchText = SearchText,
        SourceApp = SourceApp,
        SourceAppName = SourceAppName,
        SourceAppPath = SourceAppPath,
        FirstCopiedAt = FirstCopiedAt,
        LastCopiedAt = LastCopiedAt,
        CopyCount = CopyCount,
        PinnedAt = PinnedAt,
        ByteCount = ByteCount,
        CharacterCount = CharacterCount,
        LineCount = LineCount,
        IsRichText = IsRichText,
        IsTruncated = IsTruncated,
        LinkUrl = LinkUrl,
        ColorHex = ColorHex,
        FilePaths = FilePaths,
        ImagePixelSize = ImagePixelSize,
        Thumbnail = Thumbnail,
    };

    public static StoredClip FromDraft(ClipDraft draft, Guid id, DateTimeOffset now) => new()
    {
        Id = id,
        Kind = draft.Kind,
        Title = draft.Title,
        SearchText = draft.SearchText,
        ContentHash = draft.ContentHash,
        SourceApp = draft.SourceApp,
        SourceAppName = draft.SourceAppName,
        SourceAppPath = draft.SourceAppPath,
        FirstCopiedAt = now,
        LastCopiedAt = now,
        CopyCount = 1,
        PinnedAt = null,
        IsRichText = draft.IsRichText,
        IsTruncated = draft.IsTruncated,
        LineCount = draft.LineCount,
        CharacterCount = draft.CharacterCount,
        ByteCount = draft.ByteCount,
        LinkUrl = draft.LinkUrl,
        ColorHex = draft.ColorHex,
        FilePaths = draft.FilePaths,
        ImagePixelSize = draft.ImagePixelSize,
        Thumbnail = draft.Thumbnail,
    };
}

/// <summary>Persistence behind <see cref="HistoryStore"/>: SQLite in the app, a list in tests.</summary>
public interface IClipRepository
{
    IReadOnlyList<StoredClip> LoadAll();
    void Insert(StoredClip clip, IReadOnlyList<ClipContent> contents);
    void Update(StoredClip clip);
    void Delete(IReadOnlyList<Guid> ids);
    IReadOnlyList<ClipContent> Contents(Guid id);
    /// <summary>Approximate bytes on disk (for the Privacy tab).</summary>
    long StorageBytes();
}

public sealed class InMemoryClipRepository : IClipRepository
{
    private readonly Dictionary<Guid, (StoredClip Clip, IReadOnlyList<ClipContent> Contents)> _rows = [];

    public IReadOnlyList<StoredClip> LoadAll() => _rows.Values.Select(r => r.Clip).ToList();

    public void Insert(StoredClip clip, IReadOnlyList<ClipContent> contents) => _rows[clip.Id] = (clip, contents);

    public void Update(StoredClip clip)
    {
        if (_rows.TryGetValue(clip.Id, out var existing)) _rows[clip.Id] = (clip, existing.Contents);
    }

    public void Delete(IReadOnlyList<Guid> ids)
    {
        foreach (var id in ids) _rows.Remove(id);
    }

    public IReadOnlyList<ClipContent> Contents(Guid id) => _rows.TryGetValue(id, out var row) ? row.Contents : [];

    public long StorageBytes() => _rows.Values.Sum(r => (long)r.Clip.ByteCount);
}

/// <summary>What a delete leaves behind so Ctrl+Z can put the clip back exactly where it was.</summary>
public sealed record DeleteRecord(StoredClip Clip, IReadOnlyList<ClipContent> Contents);

/// <summary>Single-level undo with a time window (the toast's four seconds).</summary>
public sealed class UndoBuffer
{
    public static readonly TimeSpan Window = TimeSpan.FromSeconds(4);
    private DeleteRecord? _record;
    private DateTimeOffset _at;

    public void Remember(DeleteRecord record, DateTimeOffset now)
    {
        _record = record;
        _at = now;
    }

    /// <summary>The record if it is still inside the window; it is consumed either way.</summary>
    public DeleteRecord? Take(DateTimeOffset now)
    {
        var record = _record;
        _record = null;
        return record is not null && now - _at <= Window ? record : null;
    }

    public bool HasPending => _record is not null;
}

/// <summary>
/// The history: dedup by content hash, item cap (pinned exempt), age expiry, delete/restore, pins.
/// Rows are kept in memory newest-first; the repository is the durable copy.
/// </summary>
public sealed class HistoryStore
{
    private readonly IClipRepository _repository;
    private readonly List<StoredClip> _clips = [];

    public HistoryStore(IClipRepository repository)
    {
        _repository = repository;
    }

    public int MaxItems { get; set; } = 500;

    /// <summary>0 = never.</summary>
    public int ExpireAfterDays { get; set; }

    public int Version { get; private set; }

    public IReadOnlyList<ClipRow> Rows => _clips.Select(c => c.ToRow()).ToList();

    public IReadOnlyList<StoredClip> Clips => _clips;

    public int Count => _clips.Count;
    public int PinnedCount => _clips.Count(c => c.IsPinned);

    public void Load()
    {
        _clips.Clear();
        _clips.AddRange(_repository.LoadAll().OrderByDescending(c => c.LastCopiedAt));
        Version++;
    }

    public ClipRow? Row(Guid id) => Clip(id)?.ToRow();

    public StoredClip? Clip(Guid id) => _clips.FirstOrDefault(c => c.Id == id);

    public StoredClip? ClipByHash(string hash) => _clips.FirstOrDefault(c => c.ContentHash == hash);

    public IReadOnlyList<ClipContent> Contents(Guid id) => _repository.Contents(id);

    /// <summary>Insert a new clip or bump an identical one; returns the id of the row that now represents it.</summary>
    public Guid Ingest(ClipDraft draft, DateTimeOffset now)
    {
        var existingIndex = _clips.FindIndex(c => c.ContentHash == draft.ContentHash);
        if (existingIndex >= 0)
        {
            var existing = _clips[existingIndex];
            var bumped = existing with
            {
                LastCopiedAt = now,
                CopyCount = existing.CopyCount + 1,
                SourceApp = draft.SourceApp ?? existing.SourceApp,
                SourceAppName = draft.SourceAppName ?? existing.SourceAppName,
                SourceAppPath = draft.SourceAppPath ?? existing.SourceAppPath,
            };
            _clips.RemoveAt(existingIndex);
            _clips.Insert(0, bumped);
            _repository.Update(bumped);
            Version++;
            return bumped.Id;
        }

        var clip = StoredClip.FromDraft(draft, Guid.NewGuid(), now);
        _clips.Insert(0, clip);
        _repository.Insert(clip, draft.Contents);
        EnforceLimit();
        Version++;
        return clip.Id;
    }

    /// <summary>Nori pasted this clip: it moves to the top like a fresh copy.</summary>
    public void Touch(Guid id, DateTimeOffset now)
    {
        var index = _clips.FindIndex(c => c.Id == id);
        if (index < 0) return;
        var clip = _clips[index] with { LastCopiedAt = now, CopyCount = _clips[index].CopyCount + 1 };
        _clips.RemoveAt(index);
        _clips.Insert(0, clip);
        _repository.Update(clip);
        Version++;
    }

    public void TogglePin(Guid id, DateTimeOffset now)
    {
        var index = _clips.FindIndex(c => c.Id == id);
        if (index < 0) return;
        var clip = _clips[index] with { PinnedAt = _clips[index].IsPinned ? null : now };
        _clips[index] = clip;
        _repository.Update(clip);
        EnforceLimit();
        Version++;
    }

    public DeleteRecord? Delete(Guid id)
    {
        var index = _clips.FindIndex(c => c.Id == id);
        if (index < 0) return null;
        var clip = _clips[index];
        var contents = _repository.Contents(id);
        _clips.RemoveAt(index);
        _repository.Delete([id]);
        Version++;
        return new DeleteRecord(clip, contents);
    }

    /// <summary>Put a deleted clip back with its dates, pin and position.</summary>
    public Guid Restore(DeleteRecord record)
    {
        var clip = record.Clip;
        var insertAt = _clips.FindIndex(c => c.LastCopiedAt <= clip.LastCopiedAt);
        _clips.Insert(insertAt < 0 ? _clips.Count : insertAt, clip);
        _repository.Insert(clip, record.Contents);
        Version++;
        return clip.Id;
    }

    /// <summary>Removes unpinned (or all) clips; returns how many went.</summary>
    public int Clear(bool includingPinned)
    {
        var doomed = _clips.Where(c => includingPinned || !c.IsPinned).Select(c => c.Id).ToList();
        if (doomed.Count == 0) return 0;
        _clips.RemoveAll(c => doomed.Contains(c.Id));
        _repository.Delete(doomed);
        Version++;
        return doomed.Count;
    }

    /// <summary>Age expiry; pinned clips are exempt.</summary>
    public int ExpireOldItems(DateTimeOffset now)
    {
        if (ExpireAfterDays <= 0) return 0;
        var cutoff = now.AddDays(-ExpireAfterDays);
        var doomed = _clips.Where(c => !c.IsPinned && c.LastCopiedAt < cutoff).Select(c => c.Id).ToList();
        if (doomed.Count == 0) return 0;
        _clips.RemoveAll(c => doomed.Contains(c.Id));
        _repository.Delete(doomed);
        Version++;
        return doomed.Count;
    }

    public long StorageBytes() => _repository.StorageBytes();

    private void EnforceLimit()
    {
        var unpinned = _clips.Where(c => !c.IsPinned).OrderByDescending(c => c.LastCopiedAt).ToList();
        if (unpinned.Count <= MaxItems) return;
        var doomed = unpinned.Skip(MaxItems).Select(c => c.Id).ToList();
        _clips.RemoveAll(c => doomed.Contains(c.Id));
        _repository.Delete(doomed);
    }
}
